File: String/LC1106-ParsingABooleanExpression.cpp
class Solution {
public:
    bool parseBoolExpr(string expression) {
        stack<char> stk1; // save the logical operations
        stack<char> stk2; // save the "numbers"
        
        // Go from the beginning to the ending
        for (int i = 0; i < expression.size(); ++i) {
            // If it is a logical operation
            if (expression[i] == '|' || expression[i] == '&' || expression[i] == '!')
                stk1.push(expression[i]);
            // If it is a ")" we want to finish a calculation
            else if (expression[i] == ')') {
                // Count number of true and false
                int t = 0;
                int f = 0;
                // We need to calculate the expression within this "()"
                while (stk2.top() != '(') {
                    if (stk2.top() == 't')
                        t++;
                    if (stk2.top() == 'f')
                        f++;
                    stk2.pop();
                }
                // Pop out the '('
                stk2.pop();
                // X is the logic operator
                char x = stk1.top();
                stk1.pop();
                // Perform the logical calculation
                if (x == '!') {
                    if (t != 0)
                        stk2.push('f');
                    else
                        stk2.push('t');
                }
                else if (x == '&') {
                    if (f != 0)
                        stk2.push('f');
                    else
                        stk2.push('t');
                }
                else if (x == '|') {
                    if (t != 0)
                        stk2.push('t');
                    else
                        stk2.push('f');
                }
            }
            else
                stk2.push(expression[i]);
        }
        if (stk2.top() == 't') return true;
        else return false;
    }
};

File: String/LC1106-ParsingABooleanExpression.py
class Solution:
    def parseBoolExpr(self, expression: str) -> bool:
        stk1 = [] # to store all the logical operations
        stk2 = [] # to store all the boolean values
        
        for i in expression:
            # if we encountered a logical expression
            if i == '!' or i == '&' or i == '|':
                stk1.append(i)
                continue
            # if we encountered a ')'
            elif i == ')':
                # count for true and false occured
                t = 0
                f = 0
                while stk2[-1] != '(':
                    if stk2[-1] == 't':
                        t += 1
                    elif stk2[-1] == 'f':
                        f += 1
                    stk2.pop()
                # pop out the '('
                stk2.pop()
                # x is our logical operation
                x = stk1.pop()
                # perform the logical calculation
                if x == '!':
                    if t == 1:
                        stk2.append('f')
                    else:
                        stk2.append('t')
                elif x == '&':
                    if f != 0:
                        stk2.append('f')
                    else:
                        stk2.append('t')
                elif x == '|':
                    if t != 0:
                        stk2.append('t')
                    else:
                        stk2.append('f')
            # if we encounter other characters
            else:
                stk2.append(i)
        
        # If the input is valid, there will be one and only one element in stk2
        if stk2[0] == 't': return True
        else: return False
        
                
            
